### ISSUE TITLE
Getting 100% NoResponse from a server that does respond to cURL

### ISSUE THREAD

**Issue by SirVer** [http-rs/surf#117]
Using `surf 1.0.3` from crates.io, rustc 1.39.0 (4560ea788 2019-11-04) on Mac OS and this program:

```rust
#[async_std::main]
async fn main() {
    env_logger::init();    
    let rv = surf::get("https://www.spiegel.de/schlagzeilen/index.rss")
        .recv_bytes()
        .await;
    match rv {
        Ok(_) => (),
        Err(e) => println!("{:?}", e),
    }
}
```

I get 100% of the time `NoResponse` as error. However if I do `curl 'https://www.spiegel.de/schlagzeilen/index.rss'` on the commandline or use `reqwest` with tokio equivalently, I do get the expected content.

I tried using the `hyper-client` feature, but unfortunately that did not compile for me.

**Comment by SirVer** [comment]
Just checked on Linux and I get the same behavior with the same tool.

**Comment by SirVer** [comment]
Problem is also reproducible at head using `surf = { git = "https://github.com/http-rs/surf.git", rev = "380c450fce2053f3b77e9726a99d741e1a33e065" }`, which is the current master at time of writing.

**Comment by SirVer** [comment]
This is a run with using `RUST_LOG=debug` in the environment:

```
[2019-11-28T11:09:42Z DEBUG isahc::agent] agent waker listening on 127.0.0.1:57634
[2019-11-28T11:09:42Z DEBUG isahc::agent] agent took 1.732646ms to start up
[2019-11-28T11:09:42Z DEBUG isahc::agent] agent ready
[2019-11-28T11:09:42Z INFO  surf::middleware::logger::native] sending request
[2019-11-28T11:09:42Z DEBUG isahc::handler] initializing handler for request [id=0]
[2019-11-28T11:09:42Z DEBUG isahc::handler]   Trying 128.65.210.181...
[2019-11-28T11:09:42Z DEBUG isahc::handler] TCP_NODELAY set
[2019-11-28T11:09:42Z DEBUG isahc::handler] Connected to www.spiegel.de (128.65.210.181) port 443 (#0)
[2019-11-28T11:09:42Z DEBUG isahc::handler] ALPN, offering h2
[2019-11-28T11:09:42Z DEBUG isahc::handler] ALPN, offering http/1.1
[2019-11-28T11:09:42Z DEBUG isahc::handler] Cipher selection: ALL:!EXPORT:!EXPORT40:!EXPORT56:!aNULL:!LOW:!RC4:@STRENGTH
[2019-11-28T11:09:42Z DEBUG isahc::handler] successfully set certificate verify locations:
[2019-11-28T11:09:42Z DEBUG isahc::handler]   CAfile: /etc/ssl/cert.pem
      CApath: none
[2019-11-28T11:09:42Z DEBUG isahc::handler] TLSv1.2 (OUT), TLS handshake, Client hello (1):
[2019-11-28T11:09:42Z DEBUG isahc::handler] TLSv1.2 (IN), TLS handshake, Server hello (2):
[2019-11-28T11:09:42Z DEBUG isahc::handler] TLSv1.2 (IN), TLS handshake, Certificate (11):
[2019-11-28T11:09:42Z DEBUG isahc::handler] TLSv1.2 (IN), TLS handshake, Server key exchange (12):
[2019-11-28T11:09:42Z DEBUG isahc::handler] TLSv1.2 (IN), TLS handshake, Server finished (14):
[2019-11-28T11:09:42Z DEBUG isahc::handler] TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
[2019-11-28T11:09:42Z DEBUG isahc::handler] TLSv1.2 (OUT), TLS change cipher, Client hello (1):
[2019-11-28T11:09:42Z DEBUG isahc::handler] TLSv1.2 (OUT), TLS handshake, Finished (20):
[2019-11-28T11:09:42Z DEBUG isahc::handler] TLSv1.2 (IN), TLS change cipher, Client hello (1):
[2019-11-28T11:09:42Z DEBUG isahc::handler] TLSv1.2 (IN), TLS handshake, Finished (20):
[2019-11-28T11:09:42Z DEBUG isahc::handler] SSL connection using TLSv1.2 / ECDHE-RSA-CHACHA20-POLY1305
[2019-11-28T11:09:42Z DEBUG isahc::handler] ALPN, server did not agree to a protocol
[2019-11-28T11:09:42Z DEBUG isahc::handler] Server certificate:
[2019-11-28T11:09:42Z DEBUG isahc::handler]  subject: C=DE; ST=Hamburg; L=Hamburg; O=SPIEGEL ONLINE GmbH & Co. KG; CN=www.spiegel.de
[2019-11-28T11:09:42Z DEBUG isahc::handler]  start date: Sep 26 00:00:00 2018 GMT
[2019-11-28T11:09:42Z DEBUG isahc::handler]  expire date: Oct 25 12:00:00 2020 GMT
[2019-11-28T11:09:42Z DEBUG isahc::handler]  subjectAltName: host "www.spiegel.de" matched cert's "www.spiegel.de"
[2019-11-28T11:09:42Z DEBUG isahc::handler]  issuer: C=US; O=DigiCert Inc; OU=www.digicert.com; CN=GeoTrust RSA CA 2018
[2019-11-28T11:09:42Z DEBUG isahc::handler]  SSL certificate verify ok.
[2019-11-28T11:09:42Z DEBUG isahc::handler] TLSv1.2 (IN), TLS alert, Client hello (1):
[2019-11-28T11:09:42Z DEBUG isahc::handler] Empty reply from server
[2019-11-28T11:09:42Z DEBUG isahc::handler] Connection #0 to host www.spiegel.de left intact
[2019-11-28T11:09:42Z DEBUG isahc::handler] curl error: [52] Server returned nothing (no headers, no data)
[2019-11-28T11:09:42Z WARN  isahc::handler] request completed with error [id=AtomicCell { value: 0 }]: NoResponse: server did not send a response
[2019-11-28T11:09:42Z DEBUG isahc::agent] agent shutting down
```

**Comment by SirVer** [comment]
@yoshuawuyts @goto-bus-stop Friendly ping? This is currently a blocker for me to use `async-std` and `surf` over `tokio` and `reqwest`, I'd love to get some insights!

**Comment by tirr-c** [comment]
It seems like this is a bug of isahc. Isahc tries to send request body if the body is explicitly set, thus rejecting the request from the server.

The bug is reproducible only with isahc 0.8.1:

```rust
use isahc::prelude::*;

fn main() {
    let body = Body::reader([].as_ref());
    let request = Request::get("https://www.spiegel.de/schlagzeilen/index.rss").body(body).unwrap();
    let result = isahc::send(request);
    match result {
        Err(e) => println!("{:?}", e),
        Ok(_) => {},
    }
}
```

**Comment by SirVer** [comment]
@tirr-c I do not fully follow. Are you saying I am using something wrong if you say it is not a bug? I am also confused, because I do not send anything, I just want to get the site without any additional information being send and it seems reasonable that `surf::get` should just fetch the content?

So to paraphrase, somewhere there seems to be a clear bug here. I cannot say if it is in isahc or in surf. However, I opened an issue in isahc to make the maintainers aware.

**Comment by sagebind** [comment]
I added an Isahc-specific comment to that issue, thanks for reporting it. Arguably, this is a surf bug. Isahc will always send a request body if you give it a non-empty one; if you don't want to send a request body, then you need to set the request body to something that is "empty", like `isahc::Body::empty()` or simply `()`. I'm not exactly sure the best way to fix this in surf though, as surf's request body is a bit more abstracted away and doesn't seem to make a distinction between "no body" and "a body of length 0".

**Comment by tirr-c** [comment]
@SirVer Sorry, my comment was not clear; I checked out the HTTP spec and found out this is rather a bug of Surf. Some servers reject GET requests with body, and doing so is not wrong since its semantics is undefined.

As @sagebind commented, currently Surf sends a chunked-encoded body of length 0 for GET requests by default, and this confuses some servers. I think we shouldn't send any body when building GET request, unless explicitly set.

**Comment by SirVer** [comment]
> I think we shouldn't send any body when building GET request, unless explicitly set.

I agree with this approach. that seems to be what Curl and Reqwest are doing too.

**Comment by mmstick** [comment]
Just ran across this issue when trying to download debian packages from a server behind a Cloudflare CDN. Returns a 501 error due to not supporting a body in a get request.

**Comment by Nemo157** [comment]
(Seems like the same issue as #72 at a skim).

**Comment by mmstick** [comment]
Seems the fix to this will be a semver-breaking change to `http-client`.

**Comment by mmstick** [comment]
Nevermind. I found a solution that works without breaking semver.

**Comment by mmstick** [comment]
https://github.com/http-rs/http-client/pull/14 fixes the issue without requiring any changes in surf.

**Comment by tarkah** [comment]
Thanks, @mmstick! I was able to use your branch to fix this same issue. Although, I couldn't compile until making the following change...

```
--- a/src/isahc.rs
+++ b/src/isahc.rs
@@ -44,7 +44,7 @@ impl HttpClient for IsahcClient {
             let (parts, body) = req.into_parts();
 
             let body = if body.is_empty() {
-                Body::empty()
+                isahc::Body::empty()
             } else {
                 match body.len {
                     Some(len) => isahc::Body::reader_sized(body, len),
```

**Comment by mmstick** [comment]
Done

**Comment by Zizico2** [comment]
Should this Issue still be open? Is this solved?

**Comment by SirVer** [comment]
I can confirm that the original bug reported by me is fixed with surf `2.1.0`, so I think this can be closed.
